### ISSUE TITLE
More data passed to telemetry event

### ISSUE THREAD

**Issue by aselder** [zachallaun/req_telemetry#1]
I'd love it if there was the ability to add additional arbitrary metadata to the telemetry events.

If you look at Tesla, it's Telemetry middleware can be added like this:

```
{Tesla.Middleware.Telemetry, %{metadata: metadata}},
```

which simply adds the metadata to the telemetry events.

I use this a lot for sending tagging information that can be sent to APM tools like DataDog, etc.

**Comment by zachallaun** [comment]
Very doable! My suggestion as to how it might work:

```elixir
# Provide when attaching `ReqTelemetry`
req =
  Req.new()
  |> ReqTelemetry.attach(meta: %{key1: :value1})

# Provide when issuing request
Req.get!(req, url: "https://example.org", telemetry: [meta: %{key2: value2}])
```

I'd propose that `:meta` must be a map so that we can merge any defaults given in `attach` with any further metadata provided when the request is made.

Your telemetry event handler would then have access to it in the third argument:

```elixir
def handle_event([:req, :request, :pipeline, :start], _measurements, %{meta: meta}, _config) do
  %{key1: :value1, key2: :value2} = meta  
  # ...
end
```

One thing I'm somewhat unsure of is what `:meta` should be in `handle_event` if nothing is specified in either of `attach` or at time of request. I'm leaning towards it either being `nil` or not present at all, both of which allow you to pattern match on it (vs. an empty map).

**Comment by aselder** [comment]
I've put together a really basic PR. I defaulted `metadata` to an empty maps, but easy enough to change around.

**Comment by zachallaun** [comment]
Closed by #2
